redfish-exporter/slurm/slurm.go

```diff
@@ -50,136 +53,188 @@ type SlurmServerConfig struct {
 	if err != nil {
-		return err
+		return nil
 	}
 	defer resp.Body.Close()
 
+	log.Printf("[slurm] ping success: %v\n", resp.StatusCode)
 	return nil
 }
 
-func (c *Client) GetNodes() ([]string, error) {
-	var nodes []string
+func (c *Client) updateNodeState(nodeName, state string) error {
 	apiCall := func() (interface{}, *http.Response, error) {
 		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
-		jreq := c.apiClient.SlurmAPI.SlurmV0039GetNodes(ctx)
-		res, resp, err := c.apiClient.SlurmAPI.SlurmV0039GetNodesExecute(jreq)
-		cancel()
-		if err != nil {
-			return res, resp, err
-		} else if resp.StatusCode != 200 {
-			return res, resp, fmt.Errorf("invalid status code: %v", resp.StatusCode)
+		defer cancel()
+
+		// Step 1: Call the PostNode method using reflection
+		postNodeVals := c.helper["PostNode"].Func.Call([]reflect.Value{
+			reflect.ValueOf(c.apiClient.SlurmAPI),
+			reflect.ValueOf(ctx),
+			reflect.ValueOf(nodeName),
+		})
+
+		// Check if the call produced results
+		if len(postNodeVals) == 0 {
+			return nil, nil, fmt.Errorf("PostNode call returned no values")
 		}
-		return res, resp, nil
-	}
-
-	res, resp, err := CallWithRetry(apiCall, maxRetries, baseDelay)
-	if err != nil {
-		return nodes, err
-	}
-	defer resp.Body.Close()
-
-	log.Printf("[slurm] get nodes: %+v\n", nodes)
-	temp := res.(*slurmrestdapi.V0039NodesResponse)
-	for _, node := range temp.GetNodes() {
-		nodes = append(nodes, *node.Name)
-	}
-	return nodes, nil
-}
 
-func (c *Client) getConnectionStatus() error {
-	apiCall := func() (interface{}, *http.Response, error) {
-		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
-		jreq := c.apiClient.SlurmAPI.SlurmV0039Ping(ctx)
-		res, resp, err := c.apiClient.SlurmAPI.SlurmV0039PingExecute(jreq)
-		cancel()
-		if err != nil {
-			return res, resp, err
-		} else if resp.StatusCode != 200 {
-			return res, resp, fmt.Errorf("invalid status code: %v", resp.StatusCode)
+		// Step 2: Find and call the UpdateNodeMsg method on the request object
+		newInstance := reflect.New(postNodeVals[0].Type()).Elem()
+		instanceType := newInstance.Type()
+
+		for i := 0; i < instanceType.NumMethod(); i++ {
+			method := instanceType.Method(i)
+			if strings.Contains(method.Name, "UpdateNodeMsg") {
+				// Create a new UpdateNodeMsg request
+				updateNodeMsgReq := createUpdateNodeMsgRequest(method.Type)
+				if updateNodeMsgReq.IsValid() {
+					updateNodeMsgReq.FieldByName("State").Set(reflect.ValueOf([]string{state}))
+				}
+
+				// Step 3: Call UpdateNodeMsg with the request
+				updatedNodeVals := method.Func.Call([]reflect.Value{postNodeVals[0], updateNodeMsgReq})
+				if len(updatedNodeVals) == 0 {
+					return nil, nil, fmt.Errorf("UpdateNodeMsg call returned no values")
+				}
+
+				// Step 4: Execute the PostNode method with the updated request
+				postNodeResp := c.helper["PostNodeExecute"].Func.Call([]reflect.Value{
+					reflect.ValueOf(c.apiClient.SlurmAPI),
```

Comment:
There's just one corner case, if the input parameters of the PostNodeExecute function (or any other function that we call) change between Slurm versions that this code will not work. Such an API change would be a breaking change though and hopefully will happen in a corner case. If it happens, user would have to manually change this code to pass the new parameters.